### ISSUE TITLE
AC2 MLP Policy pdparam

### ISSUE THREAD

**Issue by sccrthlt** [openai/baselines#399]
I would like to utilize the MlpPolicy with AC2:

    class MlpPolicyCustom(object):
        def __init__(self, sess, ob_space, ac_space, nbatch, nsteps, reuse=False): #pylint: disable=W0613
            ob_shape = (nbatch,) + ob_space.shape
            actdim = ac_space.n
            X = tf.placeholder(tf.float32, ob_shape, name='Ob') #obs
            with tf.variable_scope("model", reuse=reuse):
                activ = tf.tanh
                h1 = activ(fc(X, 'pi_fc1', nh=64, init_scale=np.sqrt(2)))
                h2 = activ(fc(h1, 'pi_fc2', nh=64, init_scale=np.sqrt(2)))
                pi = fc(h2, 'pi', actdim, init_scale=0.01)
                h1 = activ(fc(X, 'vf_fc1', nh=64, init_scale=np.sqrt(2)))
                h2 = activ(fc(h1, 'vf_fc2', nh=64, init_scale=np.sqrt(2)))
                vf = fc(h2, 'vf', 1)[:,0]
                logstd = tf.get_variable(name="logstd", shape=[1, actdim],
                    initializer=tf.zeros_initializer())

            pdparam = tf.concat([pi, pi * 0.0 + logstd], axis=1)

            self.pdtype = make_pdtype(ac_space)
            self.pd = self.pdtype.pdfromflat(pdparam)

            a0 = self.pd.sample()
            neglogp0 = self.pd.neglogp(a0)
            self.initial_state = None

            def step(ob, *_args, **_kwargs):
                a, v, neglogp = sess.run([a0, vf, neglogp0], {X:ob})
                print("a:", a)
                return a, v, self.initial_state, neglogp

            def value(ob, *_args, **_kwargs):
                return sess.run(vf, {X:ob})

            self.X = X
            self.pi = pi
            self.vf = vf
            self.step = step
            self.value = value

However the following line concatenates the policy tensor along `axis=1`, changing my `ac_space` shape from `shape=(2,10)` to `shape=(2,20)`:

    pdparam = tf.concat([pi, pi * 0.0 + logstd], axis=1)

What is the purpose of concatenating the policy along this axis with `logstd`?

**Comment by pzhokhov** [comment]
The MlpPolicyCustom code that you are providing seems to work specifically with the continuous action spaces (gym.spaces.Box), in which gaussian distribution is used to generate stochastic policy (i.e. mlp network outputs means and logs of standard deviations of a gaussian distribution, and action selection is performed via sampling from this distribution; note also that in the current implementation logstd is not actually being output by neural network, rather, is being set externally). In the latest version of master branch we have updated the policy distribution API to build proper output layer for policy network given different types of action spaces, see e.g. here https://github.com/openai/baselines/blob/9cf95a00542dc87b413e765d8c0fdcf0b1684cdc/baselines/ppo2/policies.py#L135 and here https://github.com/openai/baselines/blob/9cf95a00542dc87b413e765d8c0fdcf0b1684cdc/baselines/common/distributions.py#L86

**Comment by sccrthlt** [comment]
pzhokhov thank you for your response. I should have provided more information though:

The action space I am working with is gym.spaces.Discrete.  I subclass my own custom environment with the following:

    class CustomEnvWrapper(CustomEnv):
        def __init__(self):
            self.action_space = gym.spaces.Discrete(10)
            self.observation_space = gym.spaces.Box(low=0, high=1000, shape=(23,))
            super(CustomEnvWrapper, self).__init__()

I understand how Actor Critic (AC2) is performing gradient ascent to maximize the performance function (pi). In my case this is a discrete probability distribution over the discrete action space. 

However I don't see the purpose of the `logstd`.  We are already performing softmax with logits in calculating `pgloss` within `a2c.py`:

    neglogpac = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=train_model.pi, labels=A)
    pg_loss = tf.reduce_mean(ADV * neglogpac)
